src/plugins_types/lyds_tree.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * @file lyds_tree.c
+ * @author Adam Piecek <piecek@cesnet.cz>
+ * @brief Internal type plugin for sorting data nodes.
+ *
+ * Copyright (c) 2019-2023 CESNET, z.s.p.o.
+ *
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#include "plugins_types.h"
+
+#include <assert.h> /* assert */
+#include <stddef.h> /* NULL */
+#include <string.h> /* memset */
+
+#include "common.h"
+#include "compat.h"
+#include "libyang.h"
+#include "tree_data_sorted.h"
+
+static void lyplg_type_free_lyds(const struct ly_ctx *ctx, struct lyd_value *value);
+
+static LY_ERR
+lyplg_type_store_lyds(const struct ly_ctx *ctx, const struct lysc_type *type, const void *value,
+        size_t UNUSED(value_len), uint32_t options, LY_VALUE_FORMAT format, void *UNUSED(prefix_data),
+        uint32_t UNUSED(hints), const struct lysc_node *UNUSED(ctx_node), struct lyd_value *storage,
+        struct lys_glob_unres *UNUSED(unres), struct ly_err_item **UNUSED(err))
+{
+    int ret;
+    struct rb_node *rbt = NULL;
+    struct lyd_node *leader;
+    struct lyd_value_lyds_tree *val = NULL;
+
+    if (format == LY_VALUE_CANON) {
+        /* The canonical value for lyds_tree type is the empty string, so @p value is like NULL. */
+        memset((storage)->fixed_mem, 0, LYD_VALUE_FIXED_MEM_SIZE);
+        storage->realtype = type;
+        return LY_SUCCESS;
+    } else if ((format != LY_VALUE_LYB) || (options & LYPLG_TYPE_STORE_DYNAMIC)) {
+        return LY_EVALID;
+    }
+
+    /* Prepare value memory. */
+    leader = (struct lyd_node *)value;
+    LYPLG_TYPE_VAL_INLINE_PREPARE(storage, val);
+    LY_CHECK_ERR_GOTO(!val, ret = LY_EMEM, cleanup);
+
+    /* Create a new Red-black tree. The insertion of additional data nodes should be done via lyds_insert(). */
+    ret = rb_create_node(leader, &rbt);
+    LY_CHECK_GOTO(ret, cleanup);
+
+    /* Set the root of the Red-black tree. */
+    storage->realtype = type;
+    val->rbt = rbt;
+
+cleanup:
+    if (ret) {
+        lyplg_type_free_lyds(ctx, storage);
+    }
+
+    return ret;
+}
+
+static void
+lyplg_type_free_lyds(const struct ly_ctx *UNUSED(ctx), struct lyd_value *value)
+{
+    struct lyd_value_lyds_tree *val = NULL;
+
+    /* The canonical value is not used at all. */
+    assert(!value->_canonical);
+    LYD_VALUE_GET(value, val);
+    /* Release Red-black tree. */
```

Comment:
Please try to stick to the rule of having an empty line before every comment (unless there is `{` or something similar).